File: android-native/FretboardHeatmap/app/src/main/java/com/example/fretboardheatmap/FretPositions.kt
package com.example.fretboardheatmap

/**
 * Manages fret note and finger positions
 */
object FretPositions {

    // scales map
    val scales: Map<String, Map<String, Map<Int, List<Int>>>> = mapOf(
        "Min Pentatonic" to mapOf(
            "G" to mapOf(0 to listOf(1, 3, 6, 8, 10, 13), 1 to listOf(1, 3, 6, 8, 11, 13), 2 to listOf(3, 5, 7, 10, 12), 3 to listOf(3, 5, 8, 10, 12), 4 to listOf(1, 3, 5, 8, 10, 13), 5 to listOf(1, 3, 6, 8, 10, 13)),
            "D" to mapOf(0 to listOf(1, 3, 5, 8, 10), 1 to listOf(1, 3, 6, 8, 10), 2 to listOf(2, 5, 7, 10, 12), 3 to listOf(0, 3, 5, 7, 10, 12), 4 to listOf(0, 3, 5, 8, 10, 12), 5 to listOf(1, 3, 5, 8, 10)),
            "C" to mapOf(0 to listOf(1, 3, 6, 8, 11, 13), 1 to listOf(1, 4, 6, 8, 11, 13), 2 to listOf(0, 3, 5, 8, 10, 12), 3 to listOf(1, 3, 5, 8, 10, 13), 4 to listOf(1, 3, 6, 8, 10, 13), 5 to listOf(1, 3, 6, 8, 11, 13)),
            "E" to mapOf(0 to listOf(0, 3, 5, 7, 10, 12), 1 to listOf(0, 3, 5, 8, 10, 12), 2 to listOf(0, 2, 4, 7, 9, 12), 3 to listOf(0, 2, 5, 7, 9, 12), 4 to listOf(0, 2, 5, 7, 10, 12), 5 to listOf(0, 3, 5, 7, 10, 12)),
            "A" to mapOf(0 to listOf(0, 3, 5, 8, 10, 12), 1 to listOf(1, 3, 5, 8, 10, 13), 2 to listOf(2, 5, 7, 9, 12), 3 to listOf(2, 5, 7, 10, 12), 4 to listOf(0, 3, 5, 7, 10, 12), 5 to listOf(0, 3, 5, 8, 10, 12))
        ),
        "Maj Pentatonic" to mapOf(
            "G" to mapOf(0 to listOf(0, 3, 5, 7, 10, 12), 1 to listOf(0, 3, 5, 8, 10, 12), 2 to listOf(0, 2, 4, 7, 9, 12), 3 to listOf(0, 2, 5, 7, 9, 12), 4 to listOf(0, 2, 5, 7, 10, 12), 5 to listOf(0, 3, 5, 7, 10, 12)),
            "D" to mapOf(0 to listOf(0, 2, 5, 7, 10, 12), 1 to listOf(0, 3, 5, 7, 10, 12), 2 to listOf(2, 4, 7, 9, 11), 3 to listOf(0, 2, 4, 7, 9, 12), 4 to listOf(0, 2, 5, 7, 9, 12), 5 to listOf(0, 2, 5, 7, 10, 12)),
            "C" to mapOf(0 to listOf(0, 3, 5, 8, 10, 12), 1 to listOf(1, 3, 5, 8, 10), 2 to listOf(0, 2, 5, 7, 9, 12), 3 to listOf(0, 2, 5, 7, 10, 12), 4 to listOf(3, 5, 7, 10, 12), 5 to listOf(0, 3, 5, 8, 10, 12)),
            "E" to mapOf(0 to listOf(0, 2, 4, 7, 9, 12), 1 to listOf(2, 5, 7, 9, 12), 2 to listOf(1, 4, 6, 9, 11), 3 to listOf(2, 4, 6, 9, 11), 4 to listOf(2, 4, 7, 9, 11), 5 to listOf(2, 4, 7, 9, 12)),
            "A" to mapOf(0 to listOf(2, 5, 7, 9, 12), 1 to listOf(2, 5, 7, 10, 12), 2 to listOf(2, 4, 6, 9, 11), 3 to listOf(2, 4, 7, 9, 11), 4 to listOf(2, 4, 7, 9, 12), 5 to listOf(2, 5, 7, 9, 12))
        ),
        "Ionian" to mapOf(
            "G" to mapOf(0 to listOf(2, 3, 5, 7, 8, 10, 12), 1 to listOf(1, 3, 5, 7, 8, 10, 12), 2 to listOf(2, 4, 5, 7, 9, 11, 12), 3 to listOf(2, 4, 5, 7, 9, 10, 12), 4 to listOf(2, 3, 5, 7, 9, 10, 12), 5 to listOf(2, 3, 5, 7, 8, 10, 12)),
            "D" to mapOf(0 to listOf(2, 3, 5, 7, 9, 10, 12), 1 to listOf(2, 3, 5, 7, 8, 10, 12), 2 to listOf(2, 4, 6, 7, 9, 11, 12), 3 to listOf(2, 4, 5, 7, 9, 11, 12), 4 to listOf(2, 4, 5, 7, 9, 10, 12), 5 to listOf(2, 3, 5, 7, 9, 10, 12)),
            "C" to mapOf(0 to listOf(1, 3, 5, 7, 8, 10, 12), 1 to listOf(1, 3, 5, 6, 8, 10, 12), 2 to listOf(2, 4, 5, 7, 3, 4, 12), 3 to listOf(2, 3, 5, 7, 9, 10, 12), 4 to listOf(2, 3, 5, 7, 8, 10, 12), 5 to listOf(1, 3, 5, 7, 8, 10, 12)),
            "E" to mapOf(0 to listOf(2, 4, 5, 7, 9, 11, 12), 1 to listOf(2, 4, 5, 7, 9, 10, 12), 2 to listOf(1, 2, 4, 6, 8, 9, 11), 3 to listOf(1, 2, 4, 6, 7, 9, 11), 4 to listOf(2, 4, 6, 7, 9, 11, 12), 5 to listOf(2, 4, 5, 7, 9, 11, 12)),
            "A" to mapOf(0 to listOf(2, 4, 5, 7, 9, 10, 12), 1 to listOf(2, 3, 5, 7, 9, 10, 12), 2 to listOf(1, 2, 4, 6, 7, 9, 11), 3 to listOf(2, 4, 6, 7, 9, 11, 12), 4 to listOf(2, 4, 5, 7, 9, 11, 12), 5 to listOf(2, 4, 5, 7, 9, 10, 12))
        )
    )

    // chords map
    val chords: Map<String, Map<String, Map<Int, List<Int>>>> = mapOf(
        "Major" to mapOf(
            "G" to mapOf(0 to listOf(3), 4 to listOf(2), 5 to listOf(3)),
            "D" to mapOf(0 to listOf(2), 1 to listOf(3), 2 to listOf(2)),
            "C" to mapOf(1 to listOf(1), 3 to listOf(2), 4 to listOf(3)),
            "E" to mapOf(2 to listOf(1), 3 to listOf(2), 4 to listOf(2)),
            "A" to mapOf(1 to listOf(2), 2 to listOf(2), 3 to listOf(2))
        ),
        "Minor" to mapOf(
            "A" to mapOf(1 to listOf(1), 2 to listOf(2), 3 to listOf(2)),
            "E" to mapOf(3 to listOf(2), 4 to listOf(2)),
            "D" to mapOf(0 to listOf(1), 1 to listOf(3), 2 to listOf(2)),
            "G" to mapOf(0 to listOf(3), 1 to listOf(3), 2 to listOf(3), 3 to listOf(5), 4 to listOf(5), 5 to listOf(3)),
            "C" to mapOf(0 to listOf(3), 1 to listOf(4), 2 to listOf(5), 3 to listOf(5))
        )
    )

    // finger numbers map
    val fingerNumbers: Map<String, Map<String, Map<String, String>>> = mapOf(
        "Major" to mapOf(
            "G" to mapOf("5,3" to "2", "4,2" to "1", "0,3" to "3"),
            "D" to mapOf("2,2" to "1", "1,3" to "3", "0,2" to "2"),
            "C" to mapOf("4,3" to "3", "3,2" to "2", "1,1" to "1"),
            "E" to mapOf("2,1" to "1", "3,2" to "3", "4,2" to "2"),
            "A" to mapOf("3,2" to "1", "2,2" to "2", "1,2" to "3")
        ),
        "Minor" to mapOf(
            "A" to mapOf("1,1" to "1", "2,2" to "3", "3,2" to "2"),
            "E" to mapOf("3,2" to "2", "4,2" to "1"),
            "D" to mapOf("0,1" to "1", "1,3" to "3", "2,2" to "2"),
            "C" to mapOf("3,5" to "3", "2,5" to "4", "1,4" to "2", "0,3" to "1"),
            "G" to mapOf("5,3" to "1", "4,5" to "3", "3,5" to "4", "2,3" to "1", "1,3" to "1", "0,3" to "1")
        )
    )

//    /**
//     * Prepare dictionaries for top menu dropdown
//     * @param forTopMenu top menu choice
//     */
//    fun dropdownChoices(forTopMenu: TopMenuChoice): List<String> {
//        return when (forTopMenu) {
//            TopMenuChoice.CHORDS -> chords.keys.sorted()
//            TopMenuChoice.SCALES -> scales.keys.sorted()
//        }
//    }

    /**
     * Prepare fret notes for heatmap
     * @param topMenu top menu choice
     * @param dropdownChoice dropdown option choice
     * @param root root note choice
     */
    fun getFretMap(
        topMenu: TopMenuChoice?,
        dropdownChoice: String,
        root: String
    ): Map<Int, List<Int>> {
        if (topMenu == null) return emptyMap()

        return when (topMenu) {
            TopMenuChoice.CHORDS -> chords[dropdownChoice]?.get(root) ?: emptyMap()
            TopMenuChoice.SCALES -> scales[dropdownChoice]?.get(root) ?: emptyMap()
        }
    }

    /**
     * Get finger numbers for chords
     * @param dropdownChoice dropdown option choice
     * @param root root note choice
     * @param string the current string
     * @param fret the current fret
     */
    fun getFingerNumber(dropdownChoice: String, root: String, string: Int, fret: Int): String {
        val key = "$string,$fret"
        return fingerNumbers[dropdownChoice]?.get(root)?.get(key) ?: ""
    }
}

File: ios-native/FretboardHeatmap/FretPositions.swift
//
//  FretPositions.swift
//  FretboardHeatmap
//
//  Created by Sean Avery Suguitan on 2025-12-26.
//

import Foundation

/// Available top menu choices
enum menuChoice: String {
    case chords
    case scales
}

/// Manages fret note and finger positions
struct FretPositions {
    
    /// Prepare dictionaries for top menu dropdown
    /// - Parameters:
    ///   - topMenu: Top menu choice
    /// - Returns: Options for right top menu dropdown
    static func dropdownChoices(for topMenu: menuChoice) -> [String] {
        switch topMenu {
        case .chords:
            return Array(chords.keys).sorted()
        case .scales:
            return Array(scales.keys).sorted()
        }
    }
    
    /// Prepare fret notes for heatmap
    /// - Parameters:
    ///    - topMenu: Top menu choice
    ///    - dropdownChoice: Top menu dropdown choice
    ///    - root: Root note choice
    /// - Returns: Fret positions for right dropdown and root choice
    static func getFretMap(topMenu: menuChoice?, dropdownChoice: String, root: String) -> [Int: [Int]] {
        guard let topMenu = topMenu else { return [:] }
        
        switch topMenu {
        case .chords:
            return chords[dropdownChoice]?[root] ?? [:]
        case .scales:
            return scales[dropdownChoice]?[root] ?? [:]
        }
        
    }
    
    /// Get finger numbers for chords
    /// - Parameters:
    ///    - dropdownChoice: Top menu dropdown choice
    ///    - root: Root note choice
    ///    - string: current string
    ///    - fret: current fret
    /// - Returns: finger number positions for chords
    static func getFingerNumber(dropdownChoice: String, root: String, string: Int, fret: Int) -> String {
        let key = "\(string),\(fret)"
        
        return fingerNumbers[dropdownChoice]?[root]?[key] ?? ""
        
    }
    
    // MARK: Scales dictionary
    private static let scales: [String: [String: [Int: [Int]]]] = [
        "Min Pentatonic": [
            "G": [0: [1, 3, 6, 8, 10, 13], 1: [1, 3, 6, 8, 11, 13], 2: [3, 5, 7, 10, 12], 3: [3, 5, 8, 10, 12], 4: [1, 3, 5, 8, 10, 13], 5: [1, 3, 6, 8, 10, 13]],
            "D": [0: [1, 3, 5, 8, 10], 1: [1, 3, 6, 8, 10], 2: [2, 5, 7, 10, 12], 3: [0, 3, 5, 7, 10, 12], 4: [0, 3, 5, 8, 10, 12], 5: [1, 3, 5, 8, 10]],
            "C": [0: [1, 3, 6, 8, 11, 13], 1: [1, 4, 6, 8, 11, 13], 2: [0, 3, 5, 8, 10, 12], 3: [1, 3, 5, 8, 10, 13], 4: [1, 3, 6, 8, 10, 13], 5: [1, 3, 6, 8, 11, 13]],
            "E": [0: [0, 3, 5, 7, 10, 12], 1: [0, 3, 5, 8, 10, 12], 2: [0, 2, 4, 7, 9, 12], 3: [0, 2, 5, 7, 9, 12], 4: [0, 2, 5, 7, 10, 12], 5: [0, 3, 5, 7, 10, 12]],
            "A": [0: [0, 3, 5, 8, 10, 12], 1: [1, 3, 5, 8, 10, 13], 2: [2, 5, 7, 9, 12], 3: [2, 5, 7, 10, 12], 4: [0, 3, 5, 7, 10, 12], 5: [0, 3, 5, 8, 10, 12]]
        ],
        "Maj Pentatonic": [
            "G": [0: [0, 3, 5, 7, 10, 12], 1: [0, 3, 5, 8, 10, 12], 2: [0, 2, 4, 7, 9, 12], 3: [0, 2, 5, 7, 9, 12], 4: [0, 2, 5, 7, 10, 12], 5: [0, 3, 5, 7, 10, 12]],
            "D": [0: [0, 2, 5, 7, 10, 12], 1: [0, 3, 5, 7, 10, 12], 2: [2, 4, 7, 9, 11], 3: [0, 2, 4, 7, 9, 12], 4: [0, 2, 5, 7, 9, 12], 5: [0, 2, 5, 7, 10, 12]],
            "C": [0: [0, 3, 5, 8, 10, 12], 1: [1, 3, 5, 8, 10], 2: [0, 2, 5, 7, 9, 12], 3: [0, 2, 5, 7, 10, 12], 4: [3, 5, 7, 10, 12], 5: [0, 3, 5, 8, 10, 12]],
            "E": [0: [0, 2, 4, 7, 9, 12], 1: [2, 5, 7, 9, 12], 2: [1, 4, 6, 9, 11], 3: [2, 4, 6, 9, 11], 4: [2, 4, 7, 9, 11], 5: [2, 4, 7, 9, 12]],
            "A": [0: [2, 5, 7, 9, 12], 1: [2, 5, 7, 10, 12], 2: [2, 4, 6, 9, 11], 3: [2, 4, 7, 9, 11], 4: [2, 4, 7, 9, 12], 5: [2, 5, 7, 9, 12]]
        ],
        "Ionian": [
            "G": [0: [2, 3, 5, 7, 8, 10, 12], 1: [1, 3, 5, 7, 8, 10, 12], 2: [2, 4, 5, 7, 9, 11, 12], 3: [2, 4, 5, 7, 9, 10, 12], 4: [2, 3, 5, 7, 9, 10, 12], 5: [ 2, 3, 5, 7, 8, 10, 12]],
            "D": [0: [2, 3, 5, 7, 9, 10, 12], 1: [2, 3, 5, 7, 8, 10, 12], 2: [2, 4, 6, 7, 9, 11, 12], 3: [2, 4, 5, 7, 9, 11, 12], 4: [2, 4, 5, 7, 9, 10, 12], 5: [2, 3, 5, 7, 9, 10, 12]],
            "C": [0: [1, 3, 5, 7, 8, 10, 12], 1:[1, 3, 5, 6, 8, 10, 12], 2: [2, 4, 5, 7, 3, 4, 12], 3: [2,3, 5, 7, 9, 10, 12], 4: [2, 3, 5, 7, 8, 10, 12], 5: [1, 3, 5, 7, 8, 10, 12]],
            "E": [0: [2, 4, 5, 7, 9, 11, 12], 1: [2, 4, 5, 7, 9, 10, 12], 2: [1, 2, 4, 6, 8, 9, 11], 3: [1, 2, 4, 6, 7, 9, 11], 4: [2, 4, 6, 7, 9, 11, 12], 5: [2, 4, 5, 7, 9, 11, 12]],
            "A": [0: [2, 4, 5, 7, 9, 10, 12], 1: [2, 3, 5, 7, 9, 10, 12], 2: [1, 2, 4, 6, 7, 9, 11], 3: [2, 4, 6, 7, 9, 11, 12], 4: [2, 4, 5, 7, 9, 11, 12], 5: [2, 4, 5, 7, 9, 10, 12]]
        ]
    ]
    
    // MARK: - Chords dictionary 
    private static let chords: [String: [String: [Int: [Int]]]] = [
        "Major": [
            "G": [0: [3], 4: [2], 5: [3]],
            "D": [0: [2], 1: [3], 2: [2]],
            "C": [1: [1], 3: [2], 4: [3]],
            "E": [2: [1], 3: [2], 4: [2]],
            "A": [1: [2], 2: [2], 3: [2]]
        ],
        "Minor": [
            "A": [1: [1], 2: [2], 3: [2]],
            "E": [3: [2], 4: [2]],
            "D": [0: [1], 1: [3], 2: [2]],
            "G": [0: [3], 1: [3], 2: [3], 3: [5], 4: [5], 5: [3]],
            "C": [0: [3], 1: [4], 2: [5], 3: [5]]
        ]
    ]
    
    // MARK: Finger numbers dictionary
    private static let fingerNumbers: [String: [String: [String: String]]] = [
        "Major": [
                "G": ["5,3": "2", "4,2": "1", "0,3": "3"],
                "D": ["2,2": "1", "1,3": "3", "0,2": "2"],
                "C": ["4,3": "3", "3,2": "2", "1,1": "1"],
                "E": ["2,1": "1", "3,2": "3", "4,2": "2"],
                "A": ["3,2": "1", "2,2": "2", "1,2": "3"]
            ],
        "Minor": [
            "A": ["1,1": "1", "2,2": "3", "3,2": "2"],
            "E": ["3,2": "2", "4,2": "1"],
            "D": ["0,1": "1", "1,3": "3", "2,2": "2"],
            "C": ["3,5": "3", "2,5": "4", "1,4": "2", "0,3": "1"],
            "G": ["5,3": "1", "4,5": "3", "3,5": "4", "2,3": "1", "1,3": "1", "0,3": "1"]
        ]
    ]
}
